tree/readspeed/src/ReadSpeed.cxx

```diff
@@ -0,0 +1,398 @@
+      auto nReminderClusters = nClustersInThisFile % maxTasksPerFile;
+      const auto &clustersInThisFile = *clustersIt;
+      for (auto i = 0ULL; i < nClustersInThisFile; ++i) {
+         const auto start = clustersInThisFile[i].fStart;
+         // We lump together at least nFolds clusters, therefore
+         // we need to jump ahead of nFolds-1.
+         i += (nFolds - 1);
+         // We now add a cluster if we have some reminder left
+         if (nReminderClusters > 0) {
+            i += 1U;
+            nReminderClusters--;
+         }
+         const auto end = clustersInThisFile[i].fEnd;
+         mergedClustersIt->emplace_back(EntryRange({start, end}));
+      }
+      assert(nReminderClusters == 0 && "This should never happen, cluster-merging logic is broken.");
+   }
+
+   return mergedClusters;
+}
+
+Result ReadSpeed::EvalThroughputMT(const Data &d, unsigned nThreads)
+{
+   ROOT::TThreadExecutor pool(nThreads);
+   const auto actualThreads = ROOT::GetThreadPoolSize();
+   if (actualThreads != nThreads)
+      std::cerr << "Running with " << actualThreads << " threads even though " << nThreads << " were requested.\n";
+
+   TStopwatch clsw;
+   clsw.Start();
+   const unsigned int maxTasksPerFile =
+      std::ceil(float(ROOT::TTreeProcessorMT::GetTasksPerWorkerHint() * actualThreads) / float(d.fFileNames.size()));
+
+   const auto rangesPerFile = MergeClusters(GetClusters(d), maxTasksPerFile);
+   clsw.Stop();
+
+   const size_t nranges =
+      std::accumulate(rangesPerFile.begin(), rangesPerFile.end(), 0u, [](size_t s, auto &r) { return s + r.size(); });
+   std::cout << "Total number of tasks: " << nranges << '\n';
+
+   auto treeIdx = 0;
+   std::vector<std::vector<std::string>> fileBranchNames;
+
+   std::vector<ReadSpeedRegex> regexes;
+   if (d.fUseRegex)
+      std::transform(d.fBranchNames.begin(), d.fBranchNames.end(), std::back_inserter(regexes), [](std::string text) {
+         return ReadSpeedRegex{text, std::regex(text)};
+      });
+
+   for (const auto &fName : d.fFileNames) {
+      std::vector<std::string> branchNames;
+      if (d.fUseRegex)
+         branchNames = GetMatchingBranchNames(fName, d.fTreeNames[treeIdx], regexes);
+      else
+         branchNames = d.fBranchNames;
+
+      fileBranchNames.push_back(branchNames);
+
+      if (d.fTreeNames.size() > 1)
+         ++treeIdx;
+   }
+
+   // for each file, for each range, spawn a reading task
+   auto sumBytes = [](const std::vector<ByteData> &bytesData) -> ByteData {
+      const auto uncompressedBytes =
+         std::accumulate(bytesData.begin(), bytesData.end(), 0ull,
+                         [](ULong64_t sum, const ByteData &o) { return sum + o.fUncompressedBytesRead; });
+      const auto compressedBytes =
+         std::accumulate(bytesData.begin(), bytesData.end(), 0ull,
+                         [](ULong64_t sum, const ByteData &o) { return sum + o.fCompressedBytesRead; });
+
+      return {uncompressedBytes, compressedBytes};
+   };
```

Comment:
the body of `EvalThroughputMT` is pretty long (functions should ideally fit in one screen), so it would be good to factor this helper function out